File: rascaline-torch/tests/calculator.cpp
#include <torch/torch.h>

#include <rascaline.hpp>
#include <metatensor/torch.hpp>

#include "metatensor/torch/labels.hpp"
#include "rascaline/torch.hpp"

#include <catch.hpp>

using namespace rascaline_torch;
using namespace metatensor_torch;

static metatensor_torch::System test_system(bool positions_grad, bool cell_grad);


TEST_CASE("Calculator") {
    const auto* HYPERS_JSON = R"({
        "cutoff": 3.0,
        "delta": 4,
        "name": "bar"
    })";
    auto calculator = CalculatorHolder("dummy_calculator", HYPERS_JSON);

    SECTION("Parameters") {
        CHECK(calculator.name() == "dummy test calculator with cutoff: 3 - delta: 4 - name: bar");
        CHECK(calculator.parameters() == HYPERS_JSON);
        CHECK(calculator.cutoffs() == std::vector<double>{3.0});
    }

    SECTION("Compute -- no gradients") {
        auto system = test_system(false, false);
        auto descriptor = calculator.compute({system});

        CHECK(*descriptor->keys() == metatensor::Labels(
            {"center_type"},
            {{1}, {6}}
        ));

        // H block
        auto block = TensorMapHolder::block_by_id(descriptor, 0);
        CHECK(*block->samples() == metatensor::Labels(
            {"system", "atom"},
            {{0, 1}, {0, 2}, {0, 3}}
        ));
        CHECK(*block->properties() == metatensor::Labels(
            {"index_delta", "x_y_z"},
            {{1, 0}, {0, 1}}
        ));

        auto expected = torch::tensor({5.0, 9.0, 6.0, 18.0, 7.0, 15.0}).reshape({3, 2});
        auto values = block->values();
        CHECK(torch::all(values == expected).item<bool>());

        CHECK(values.requires_grad() == false);
        CHECK(values.grad_fn() == nullptr);

        // no gradient requested
        CHECK(block->gradients_list().empty());

        // C block
        block = TensorMapHolder::block_by_id(descriptor, 1);
        CHECK(*block->samples() == metatensor::Labels(
            {"system", "atom"},
            {{0, 0}}
        ));
        CHECK(*block->properties() == metatensor::Labels(
            {"index_delta", "x_y_z"},
            {{1, 0}, {0, 1}}
        ));

        expected = torch::tensor({4.0, 3.0}).reshape({1, 2});
        values = block->values();
        CHECK(torch::all(values == expected).item<bool>());

        CHECK(values.requires_grad() == false);
        CHECK(values.grad_fn() == nullptr);
        CHECK(block->gradients_list().empty());
    }

    SECTION("keys selection") {
        auto system = test_system(false, false);

        auto options = torch::make_intrusive<CalculatorOptionsHolder>();
        options->set_selected_keys(LabelsHolder::create({"center_type"}, {{12}, {1}}));
        auto descriptor = calculator.compute({system}, options);

        CHECK(*descriptor->keys() == metatensor::Labels(
            {"center_type"},
            {{12}, {1}}
        ));

        // empty block
        auto block = TensorMapHolder::block_by_id(descriptor, 0);
        auto values = block->values();
        CHECK(values.sizes() == std::vector<int64_t>{0, 2});

        // H block
        block = TensorMapHolder::block_by_id(descriptor, 1);
        auto expected = torch::tensor({5.0, 9.0, 6.0, 18.0, 7.0, 15.0}).reshape({3, 2});
        values = block->values();
        CHECK(torch::all(values == expected).item<bool>());
    }

    SECTION("sample selection") {
        auto system = test_system(false, false);

        auto options = torch::make_intrusive<CalculatorOptionsHolder>();
        options->set_selected_samples(LabelsHolder::create({"atom"}, {{0}, {2}}));
        auto descriptor = calculator.compute({system}, options);

        // H block
        auto block = TensorMapHolder::block_by_id(descriptor, 0);
        CHECK(*block->samples() == metatensor::Labels(
            {"system", "atom"},
            {{0, 2}}
        ));

        // C block
        block = TensorMapHolder::block_by_id(descriptor, 1);
        CHECK(*block->samples() == metatensor::Labels(
            {"system", "atom"},
            {{0, 0}}
        ));
    }

    SECTION("properties selection") {
        auto system = test_system(false, false);

        auto options = torch::make_intrusive<CalculatorOptionsHolder>();
        options->set_selected_properties(LabelsHolder::create({"index_delta"}, {{1}, {12}}));
        auto descriptor = calculator.compute({system}, options);

        // H block
        auto block = TensorMapHolder::block_by_id(descriptor, 0);
        CHECK(*block->properties() == metatensor::Labels(
            {"index_delta", "x_y_z"},
            {{1, 0}}
        ));

        // C block
        block = TensorMapHolder::block_by_id(descriptor, 1);
        CHECK(*block->properties() == metatensor::Labels(
            {"index_delta", "x_y_z"},
            {{1, 0}}
        ));
    }


    SECTION("Compute -- all gradients") {
        auto system = test_system(true, false);

        auto options = torch::make_intrusive<CalculatorOptionsHolder>();
        options->gradients.emplace_back("positions");
        auto descriptor = calculator.compute({system}, options);

        CHECK(*descriptor->keys() == metatensor::Labels(
            {"center_type"},
            {{1}, {6}}
        ));

        // H block
        auto block = TensorMapHolder::block_by_id(descriptor, 0);

        auto values = block->values();
        CHECK(values.requires_grad() == true);

        auto grad_fn = values.grad_fn();
        REQUIRE(grad_fn);
        CHECK_THAT(grad_fn->name(), Catch::Matchers::Contains("rascaline_torch::RascalineAutograd"));

        // forward gradients
        auto gradient = TensorBlockHolder::gradient(block, "positions");
        CHECK(*gradient->samples() == metatensor::Labels(
            {"sample", "system", "atom"},
            {
                {0, 0, 0}, {0, 0, 1}, {0, 0, 2},
                {1, 0, 1}, {1, 0, 2}, {1, 0, 3},
                {2, 0, 2}, {2, 0, 3},
            }
        ));
        auto expected = torch::tensor({
            0.0, 1.0, /**/ 0.0, 1.0, /**/ 0.0, 1.0,
            0.0, 1.0, /**/ 0.0, 1.0, /**/ 0.0, 1.0,
            0.0, 1.0, /**/ 0.0, 1.0, /**/ 0.0, 1.0,
            0.0, 1.0, /**/ 0.0, 1.0, /**/ 0.0, 1.0,
            0.0, 1.0, /**/ 0.0, 1.0, /**/ 0.0, 1.0,
            0.0, 1.0, /**/ 0.0, 1.0, /**/ 0.0, 1.0,
            0.0, 1.0, /**/ 0.0, 1.0, /**/ 0.0, 1.0,
            0.0, 1.0, /**/ 0.0, 1.0, /**/ 0.0, 1.0,
        }).reshape({8, 3, 2});
        CHECK(torch::all(gradient->values() == expected).item<bool>());

        // C block
        block = TensorMapHolder::block_by_id(descriptor, 1);

        values = block->values();
        CHECK(values.requires_grad() == true);

        grad_fn = values.grad_fn();
        REQUIRE(grad_fn);
        CHECK_THAT(grad_fn->name(), Catch::Matchers::Contains("rascaline_torch::RascalineAutograd"));

        // forward gradients
        gradient = TensorBlockHolder::gradient(block, "positions");
        CHECK(*gradient->samples() == metatensor::Labels(
            {"sample", "system", "atom"},
            {{0, 0, 0}, {0, 0, 1}}
        ));
        expected = torch::tensor({
            0.0, 1.0, /**/ 0.0, 1.0, /**/ 0.0, 1.0,
            0.0, 1.0, /**/ 0.0, 1.0, /**/ 0.0, 1.0,
        }).reshape({2, 3, 2});
        CHECK(torch::all(gradient->values() == expected).item<bool>());
    }

    SECTION("Compute -- no forward gradients") {
        auto system = test_system(true, false);
        auto descriptor = calculator.compute({system}, /* forward_gradients */ {});

        CHECK(*descriptor->keys() == metatensor::Labels(
            {"center_type"},
            {{1}, {6}}
        ));

        // H block
        auto block = TensorMapHolder::block_by_id(descriptor, 0);

        auto values = block->values();
        CHECK(values.requires_grad() == true);

        auto grad_fn = values.grad_fn();
        REQUIRE(grad_fn);
        CHECK_THAT(grad_fn->name(), Catch::Matchers::Contains("rascaline_torch::RascalineAutograd"));

        // no forward gradients
        CHECK(block->gradients_list().empty());

        // C block
        block = TensorMapHolder::block_by_id(descriptor, 1);

        values = block->values();
        CHECK(values.requires_grad() == true);

        grad_fn = values.grad_fn();
        REQUIRE(grad_fn);
        CHECK_THAT(grad_fn->name(), Catch::Matchers::Contains("rascaline_torch::RascalineAutograd"));

        // no forward gradients
        CHECK(block->gradients_list().empty());
    }

    SECTION("Compute -- no backward gradients") {
        auto system = test_system(false, false);

        auto options = torch::make_intrusive<CalculatorOptionsHolder>();
        options->gradients.emplace_back("positions");
        auto descriptor = calculator.compute({system}, options);

        CHECK(*descriptor->keys() == metatensor::Labels(
            {"center_type"},
            {{1}, {6}}
        ));

        // H block
        auto block = TensorMapHolder::block_by_id(descriptor, 0);

        auto values = block->values();
        CHECK(values.requires_grad() == false);
        CHECK(values.grad_fn() == nullptr);

        // forward gradients
        auto gradient = TensorBlockHolder::gradient(block, "positions");
        CHECK(gradient->samples()->count() == 8);

        // C block
        block = TensorMapHolder::block_by_id(descriptor, 1);

        values = block->values();
        CHECK(values.requires_grad() == false);
        CHECK(values.grad_fn() == nullptr);

        // forward gradients
        gradient = TensorBlockHolder::gradient(block, "positions");
        CHECK(gradient->samples()->count() == 2);
    }
}

metatensor_torch::System test_system(bool positions_grad, bool cell_grad) {
    auto types = torch::tensor({6, 1, 1, 1});
    auto positions = torch::tensor({
        0.0, 0.0, 0.0,
        1.0, 1.0, 1.0,
        2.0, 2.0, 2.0,
        3.0, 3.0, 3.0,
    }).reshape({4, 3});
    positions.requires_grad_(positions_grad);

    auto cell = 10 * torch::eye(3);
    cell.requires_grad_(cell_grad);

    auto pbc = torch::ones(3, torch::TensorOptions().dtype(torch::kBool));

    return torch::make_intrusive<metatensor_torch::SystemHolder>(types, positions, cell, pbc);
}

File: python/featomic_torch/tests/calculator.py
import os
from typing import List, Optional, Union

import pytest
import torch
from metatensor.torch import Labels, TensorMap
from metatensor.torch.atomistic import System

from featomic.torch import CalculatorModule
from featomic.torch.calculators import DummyCalculator


@pytest.fixture
def system():
    return System(
        types=torch.tensor([1, 1, 8, 8]),
        positions=torch.tensor([[0.0, 0, 0], [0, 0, 1], [0, 0, 2], [0, 0, 3]]),
        cell=torch.tensor([[10.0, 0, 0], [0, 10, 0], [0, 0, 10]]),
        pbc=torch.tensor([True, True, True]),
    )


def test_parameters():
    calculator = DummyCalculator(cutoff=3.2, delta=2, name="foo")

    expected = "dummy test calculator with cutoff: 3.2 - delta: 2 - name: foo"
    assert calculator.name == expected
    assert calculator.c_name == "dummy_calculator"

    expected = '{"cutoff": 3.2, "delta": 2, "name": "foo"}'
    assert calculator.parameters == expected
    assert calculator.cutoffs == [3.2]


def test_compute(system):
    calculator = DummyCalculator(cutoff=3.2, delta=2, name="")
    descriptor = calculator.compute(system, gradients=["positions"])

    assert len(descriptor.keys) == 2
    assert descriptor.keys.names == ["center_type"]
    assert torch.all(descriptor.keys.values == torch.tensor([[1], [8]]))

    H_block = descriptor.block({"center_type": 1})
    assert H_block.values.shape == (2, 2)
    assert torch.all(H_block.values[0] == torch.tensor([2, 6]))
    assert torch.all(H_block.values[1] == torch.tensor([3, 6]))

    assert len(H_block.samples) == 2
    assert H_block.samples.names == ["system", "atom"]
    assert tuple(H_block.samples[0]) == (0, 0)
    assert tuple(H_block.samples[1]) == (0, 1)

    assert len(H_block.components) == 0

    assert len(H_block.properties) == 2
    assert H_block.properties.names == ["index_delta", "x_y_z"]
    assert tuple(H_block.properties[0]) == (1, 0)
    assert tuple(H_block.properties[1]) == (0, 1)

    gradient = H_block.gradient("positions")
    assert gradient.values.shape == (8, 3, 2)
    for i in range(gradient.values.shape[0]):
        assert torch.all(gradient.values[i, 0, :] == torch.tensor([0, 1]))
        assert torch.all(gradient.values[i, 1, :] == torch.tensor([0, 1]))
        assert torch.all(gradient.values[i, 2, :] == torch.tensor([0, 1]))

    assert len(gradient.samples) == 8
    assert gradient.samples.names == ["sample", "system", "atom"]
    assert tuple(gradient.samples[0]) == (0, 0, 0)
    assert tuple(gradient.samples[1]) == (0, 0, 1)
    assert tuple(gradient.samples[2]) == (0, 0, 2)
    assert tuple(gradient.samples[3]) == (0, 0, 3)
    assert tuple(gradient.samples[4]) == (1, 0, 0)
    assert tuple(gradient.samples[5]) == (1, 0, 1)
    assert tuple(gradient.samples[6]) == (1, 0, 2)
    assert tuple(gradient.samples[7]) == (1, 0, 3)

    assert len(gradient.components) == 1
    component = gradient.components[0]
    assert len(component) == 3
    assert component.names == ["xyz"]
    assert tuple(component[0]) == (0,)
    assert tuple(component[1]) == (1,)
    assert tuple(component[2]) == (2,)

    assert len(gradient.properties) == 2
    assert gradient.properties.names == ["index_delta", "x_y_z"]
    assert tuple(gradient.properties[0]) == (1, 0)
    assert tuple(gradient.properties[1]) == (0, 1)

    O_block = descriptor.block({"center_type": 8})
    assert O_block.values.shape == (2, 2)
    assert torch.all(O_block.values[0] == torch.tensor([4, 6]))
    assert torch.all(O_block.values[1] == torch.tensor([5, 6]))


def test_compute_native_system_error_raise(system):
    calculator = DummyCalculator(cutoff=3.2, delta=2, name="")
    with pytest.raises(ValueError, match="only `use_native_system=True` is supported"):
        calculator.compute(system, gradients=["positions"], use_native_system=False)


def test_compute_multiple_systems(system):
    systems = [system, system, system]
    calculator = DummyCalculator(cutoff=3.2, delta=2, name="")
    descriptor = calculator.compute(systems)

    H_block = descriptor.block({"center_type": 1})
    assert H_block.values.shape == (6, 2)
    expected = torch.tensor([(0, 0), (0, 1), (1, 0), (1, 1), (2, 0), (2, 1)])
    assert torch.all(H_block.samples.values == expected)

    O_block = descriptor.block({"center_type": 8})
    assert O_block.values.shape == (6, 2)


def test_key_selection(system):
    calculator = DummyCalculator(cutoff=3.2, delta=2, name="")
    descriptor = calculator.compute(
        system,
        selected_keys=Labels("center_type", torch.IntTensor([[12], [1]])),
    )
    assert torch.all(descriptor.keys.values == torch.tensor([[12], [1]]))

    H_block = descriptor.block({"center_type": 1})
    assert H_block.values.shape == (2, 2)

    missing_block = descriptor.block({"center_type": 12})
    assert missing_block.values.shape == (0, 2)


def test_samples_selection(system):
    calculator = DummyCalculator(cutoff=3.2, delta=2, name="")
    descriptor = calculator.compute(
        system,
        selected_samples=Labels("atom", torch.IntTensor([[0], [2]])),
    )

    H_block = descriptor.block({"center_type": 1})
    assert torch.all(H_block.samples.values == torch.tensor([(0, 0)]))

    O_block = descriptor.block({"center_type": 8})
    assert torch.all(O_block.samples.values == torch.tensor([(0, 2)]))


def test_properties_selection(system):
    calculator = DummyCalculator(cutoff=3.2, delta=2, name="")
    descriptor = calculator.compute(
        system,
        selected_properties=Labels("index_delta", torch.IntTensor([[0]])),
    )

    H_block = descriptor.block({"center_type": 1})
    assert torch.all(H_block.properties.values == torch.tensor([(0, 1)]))

    O_block = descriptor.block({"center_type": 8})
    assert torch.all(O_block.properties.values == torch.tensor([(0, 1)]))


def test_base_classes():
    assert DummyCalculator.__module__ == "featomic.torch.calculators"

    assert DummyCalculator.__bases__ == (CalculatorModule,)
    assert CalculatorModule.__bases__ == (torch.nn.Module,)


def test_different_device_dtype_errors(system):
    calculator = DummyCalculator(cutoff=3.2, delta=2, name="")

    message = "all systems should have the same dtype"
    with pytest.raises(TypeError, match=message):
        calculator.compute(
            [
                system.to(dtype=torch.float32),
                system.to(dtype=torch.float64),
            ]
        )

    message = "featomic only supports float64 and float32 data"
    with pytest.raises(TypeError, match=message):
        calculator.compute(system.to(dtype=torch.float16))

    # Different devices
    custom_device = None
    if can_use_mps_backend():
        custom_device = torch.device("mps:0")

    if torch.cuda.is_available():
        custom_device = torch.device("cuda:0")

    if custom_device is not None:
        device_system = system.to(device=custom_device)

        torch.set_warn_always(True)
        message = (
            "Systems data is on device .* but featomic only supports calculations "
            "on CPU. All the data will be moved to CPU and then back on device on "
            "your behalf"
        )
        with pytest.warns(match=message):
            calculator.compute(device_system)

        message = "all systems should have the same device"
        with pytest.raises(TypeError, match=message):
            calculator.compute([device_system, system])


def test_script(tmpdir):
    class TestModule(torch.nn.Module):
        def __init__(self):
            super().__init__()
            self.calculator = DummyCalculator(cutoff=3.2, delta=6, name="foo")

        def forward(
            self,
            systems: List[System],
            gradients: Optional[List[str]] = None,
            use_native_system: bool = True,
            selected_samples: Optional[Union[Labels, TensorMap]] = None,
            selected_properties: Optional[Union[Labels, TensorMap]] = None,
            selected_keys: Optional[Labels] = None,
        ) -> TensorMap:
            return self.calculator(
                systems=systems,
                gradients=gradients,
                use_native_system=use_native_system,
                selected_samples=selected_samples,
                selected_properties=selected_properties,
                selected_keys=selected_keys,
            )

    module = TestModule()
    module = torch.jit.script(module)

    with tmpdir.as_cwd():
        torch.jit.save(module, "test-save.torch")
        module = torch.jit.load("test-save.torch")


def can_use_mps_backend():
    return (
        # Github Actions M1 runners don't have a GPU accessible
        os.environ.get("GITHUB_ACTIONS") is None
        and hasattr(torch.backends, "mps")
        and torch.backends.mps.is_built()
        and torch.backends.mps.is_available()
    )
